File: web/polish.js
/* ---------------------------------------------------------------------------
   MASK POLISH — the same soft mask, steadier.

   EdgeTAM's per-frame masks are good outlines and bad neighbours: the edge
   wobbles by a pixel or two between frames, small holes open and close, and a
   subject that is standing still still shimmers. Averaging a few frames of the
   soft mask together fixes all of that at once... and destroys anything that
   moves fast, because averaging a ball across five frames IS a streak. The
   naive version of this smeared the free kick into a comet.

   So the window is not fixed. Per frame, per subject, this measures how far the
   mask's centroid moved relative to how big the mask is -- displacement over
   sqrt(area) -- and closes the temporal window as that ratio grows. A player
   drifting two pixels a frame against a 140 px body gets the whole window; a
   ball crossing thirty pixels a frame at 36 px across gets none of it, frame by
   frame, with no threshold to tune and nothing to label by hand.

   Three stages, all scaled by one 0-100 strength:

     TEMPORAL   weighted mean of the soft masks in the window. Weight is
                triangular in distance and multiplied by a motion gate that
                falls to zero once the cumulative centroid walk reaches
                ALPHA x sqrt(area). Fast things keep their own frame.
     MORPH      grayscale close then open, square structuring element. Closes
                the pinholes the tracker leaves inside a body, then removes the
                specks it leaves outside one.
     BLUR       one or two passes of a separable [1 2 1]/4. Softens the stair
                stepping the 192x192 mask upsample leaves behind.

   PARITY. server/polish.py is a numpy transcription of this file, operation for
   operation and rounding for rounding, and server/parity.py gates the two
   against each other on a synthetic sequence. Keeping them identical is what
   lets the preview in the tab and the render on the server show the same
   subject: the browser polishes mask bitmaps, the server polishes mask PNGs,
   and the arithmetic in between is the same arithmetic.

   LOADING. Like web/dither.js this is a classic script: it publishes itself on
   globalThis.MaskPolish and, under Node, on module.exports.
--------------------------------------------------------------------------- */
'use strict';

const MaskPolish = (() => {

/* How far the centroid may walk, as a fraction of the subject's own size,
 * before a neighbouring frame counts for nothing. 0.35 of sqrt(area) is about
 * a third of a body width — generous for a walking figure, hopeless for a
 * struck ball, which is the entire point. */
const ALPHA = 0.35;

const clamp = (v, a, b) => (v < a ? a : v > b ? b : v);
/* floor(v + 0.5): JS Math.round's rule, written out so Python can copy it
 * rather than inherit numpy's round-half-to-even. */
const round0 = (v) => Math.floor(v + 0.5);

/**
 * What a strength means. Everything the algorithm does is one of these three
 * integers, so "strength 70" is a reproducible thing rather than a feeling.
 *   radius  frames each side considered   0..3
 *   morph   close/open radius in pixels   0..2
 *   blur    [1 2 1]/4 passes              0..2
 */
function params(strength) {
  const u = clamp((+strength || 0) / 100, 0, 1);
  return { u, radius: round0(u * 3), morph: round0(u * 2), blur: round0(u * 2),
           alpha: ALPHA };
}

/** area (soft >= 0.5) and its centroid, in pixels. */
function stats(m, w, h) {
  let n = 0, sx = 0, sy = 0;
  for (let y = 0; y < h; y++) {
    const row = y * w;
    for (let x = 0; x < w; x++) {
      if (m[row + x] >= 0.5) { n++; sx += x; sy += y; }
    }
  }
  return n ? { area: n, cx: sx / n, cy: sy / n } : { area: 0, cx: 0, cy: 0 };
}

/**
 * The per-frame weights the temporal stage would use.
 *   win     stats() for each frame of the window, in frame order
 *   c       index of the frame being polished
 *   radius  how many frames each side to consider
 * Returns a Float64Array the same length as `win`. Exposed because it is what
 * the motion-awareness claim is actually about, and a verifier should be able
 * to read it rather than infer it from pixels.
 */
function weights(win, c, radius) {
  const w = new Float64Array(win.length);
  w[c] = 1;
  const size = Math.sqrt(win[c].area);
  if (radius <= 0 || size < 1) return w;
  const tol = ALPHA * size;
  for (let dir = -1; dir <= 1; dir += 2) {
    let disp = 0;
    for (let d = 1; d <= radius; d++) {
      const j = c + dir * d, prev = j - dir;
      if (j < 0 || j >= win.length) break;
      if (win[j].area <= 0 || win[prev].area <= 0) break;
      const dx = win[j].cx - win[prev].cx, dy = win[j].cy - win[prev].cy;
      disp += Math.sqrt(dx * dx + dy * dy);
      const tri = (radius + 1 - d) / (radius + 1);
      const gate = clamp(1 - disp / tol, 0, 1);
      const wj = tri * gate;
      if (wj <= 0) break;           // and everything past it is further still
      w[j] = wj;
    }
  }
  return w;
}

/* ------------------------------------------------------- spatial helpers
 * Separable min/max over a (2r+1)-square, edges clamped. Written the naive way
 * on purpose: r is at most 2, and a running-extremum version would be a second
 * implementation for Python to have to match exactly. */
function _extreme(src, w, h, r, max) {
  const tmp = new Float32Array(w * h), out = new Float32Array(w * h);
  for (let y = 0; y < h; y++) {
    const row = y * w;
    for (let x = 0; x < w; x++) {
      let v = src[row + x];
      for (let k = -r; k <= r; k++) {
        const xx = clamp(x + k, 0, w - 1), s = src[row + xx];
        if (max ? s > v : s < v) v = s;
      }
      tmp[row + x] = v;
    }
  }
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      let v = tmp[y * w + x];
      for (let k = -r; k <= r; k++) {
        const yy = clamp(y + k, 0, h - 1), s = tmp[yy * w + x];
        if (max ? s > v : s < v) v = s;
      }
      out[y * w + x] = v;
    }
  }
  return out;
}
const dilate = (a, w, h, r) => (r > 0 ? _extreme(a, w, h, r, true) : a);
const erode = (a, w, h, r) => (r > 0 ? _extreme(a, w, h, r, false) : a);
const close = (a, w, h, r) => erode(dilate(a, w, h, r), w, h, r);
const open = (a, w, h, r) => dilate(erode(a, w, h, r), w, h, r);

/** One pass of a separable [1 2 1]/4, edges clamped. */
function blur121(src, w, h) {
  const tmp = new Float32Array(w * h), out = new Float32Array(w * h);
  for (let y = 0; y < h; y++) {
    const row = y * w;
    for (let x = 0; x < w; x++) {
      const a = src[row + clamp(x - 1, 0, w - 1)];
      const b = src[row + x];
      const c = src[row + clamp(x + 1, 0, w - 1)];
      tmp[row + x] = (a + 2 * b + c) / 4;
    }
  }
  for (let y = 0; y < h; y++) {
    const up = clamp(y - 1, 0, h - 1) * w, dn = clamp(y + 1, 0, h - 1) * w;
    for (let x = 0; x < w; x++) {
      out[y * w + x] = (tmp[up + x] + 2 * tmp[y * w + x] + tmp[dn + x]) / 4;
    }
  }
  return out;
}

/**
 * Polish one frame's mask.
 *   win       contiguous soft masks (Float32Array, w*h, 0..1) in frame order
 *   c         index within `win` of the frame being polished
 *   strength  0..100; 0 hands the frame straight back
 *   st        optional pre-computed stats() per window frame
 * Returns a new Float32Array.
 */
function polishFrame(win, c, w, h, strength, st) {
  const p = params(strength);
  const mid = win[c];
  if (p.u <= 0) return Float32Array.from(mid);
  let cur;
  if (p.radius > 0 && win.length > 1) {
    const S = st || win.map((m) => stats(m, w, h));
    const wt = weights(S, c, p.radius);
    let sum = 0;
    for (let j = 0; j < wt.length; j++) sum += wt[j];
    if (sum > 0 && wt.some((v, j) => j !== c && v > 0)) {
      const acc = new Float32Array(w * h);
      for (let j = 0; j < wt.length; j++) {
        const wj = wt[j];
        if (wj <= 0) continue;
        const m = win[j];
        for (let q = 0; q < acc.length; q++) acc[q] += wj * m[q];
      }
      for (let q = 0; q < acc.length; q++) acc[q] = acc[q] / sum;
      cur = acc;
    } else {
      cur = Float32Array.from(mid);
    }
  } else {
    cur = Float32Array.from(mid);
  }
  if (p.morph > 0) cur = open(close(cur, w, h, p.morph), w, h, p.morph);
  for (let i = 0; i < p.blur; i++) cur = blur121(cur, w, h);
  return cur;
}

/**
 * Polish a whole sequence. The browser does not use this — it polishes the one
 * frame it is drawing — but the parity gate and any batch job do.
 */
function polishSequence(masks, w, h, strength) {
  const p = params(strength);
  if (p.u <= 0) return masks.map((m) => Float32Array.from(m));
  const st = masks.map((m) => stats(m, w, h));
  const out = [];
  for (let i = 0; i < masks.length; i++) {
    const lo = Math.max(0, i - p.radius), hi = Math.min(masks.length - 1, i + p.radius);
    out.push(polishFrame(masks.slice(lo, hi + 1), i - lo, w, h, strength,
                         st.slice(lo, hi + 1)));
  }
  return out;
}

/** 0..1 floats -> the 0..255 the server's polished PNG holds, so the two
 *  engines quantise at the same place as well as computing the same numbers. */
function quantise(m) {
  const out = new Uint8Array(m.length);
  for (let q = 0; q < m.length; q++) {
    out[q] = clamp(round0(m[q] * 255), 0, 255);
  }
  return out;
}

return { params, stats, weights, polishFrame, polishSequence, quantise,
         dilate, erode, close, open, blur121, ALPHA, version: 1 };
})();

globalThis.MaskPolish = MaskPolish;
if (typeof module !== 'undefined') module.exports = MaskPolish;

File: server/polish.py
#!/usr/bin/env python3
"""Mask polish, server side — a numpy transcription of web/polish.js.

The algorithm, the reasoning behind the motion gate and the meaning of every
number live in that file's header; this one exists so the render path can do
the same thing to a directory of mask PNGs, and so server/parity.py can prove
the two agree.

Parity is not a coincidence, it is a constraint:

  * every stage rounds to float32 exactly where the JS writes into a
    Float32Array, and computes in float64 exactly where JS arithmetic does
    (which is everywhere: JS has no float32 arithmetic, only float32 storage);
  * the term order inside the blur is the same as the JS term order;
  * round-half-up (floor(v + 0.5)) instead of numpy's round-half-to-even.

    polish_sequence(masks, strength)      list of HxW float32 -> same
    polish_dir(src, dst, strength)        mask PNGs -> polished mask PNGs
    polished_dir(job_dir, oid, strength)  the cached directory for one subject
"""
import math
import os
import shutil

import numpy as np
from PIL import Image

ALPHA = 0.35


def _round0(v):
    """floor(v + 0.5) — JavaScript's Math.round, not numpy's round-half-even."""
    return int(math.floor(v + 0.5))


def params(strength):
    u = min(1.0, max(0.0, float(strength or 0) / 100.0))
    return dict(u=u, radius=_round0(u * 3), morph=_round0(u * 2),
                blur=_round0(u * 2), alpha=ALPHA)


def load_mask(path, w=None, h=None):
    """A soft mask PNG as float32 0..1, quantised the way the browser's
    bitmapAlpha quantises an ImageBitmap: uint8 / 255 in double, then float32."""
    im = Image.open(path).convert('L')
    if w and h and im.size != (w, h):
        im = im.resize((w, h))
    return (np.asarray(im, np.float64) / 255.0).astype(np.float32)


def save_mask(m, path):
    """float32 0..1 -> the 0..255 L PNG the tracker itself writes."""
    a = np.clip(np.floor(m.astype(np.float64) * 255.0 + 0.5), 0, 255).astype(np.uint8)
    Image.fromarray(a, mode='L').save(path)


def stats(m):
    """area (soft >= 0.5) and its centroid."""
    b = m >= 0.5
    n = int(b.sum())
    if not n:
        return (0, 0.0, 0.0)
    ys, xs = np.nonzero(b)
    return (n, float(xs.sum()) / n, float(ys.sum()) / n)


def weights(win, c, radius):
    """Per-frame temporal weights — the motion gate, in the open.

    `win` is [(area, cx, cy), ...] for a contiguous run of frames, `c` the index
    of the one being polished.
    """
    w = np.zeros(len(win), np.float64)
    w[c] = 1.0
    size = math.sqrt(win[c][0])
    if radius <= 0 or size < 1:
        return w
    tol = ALPHA * size
    for direction in (-1, 1):
        disp = 0.0
        for d in range(1, radius + 1):
            j = c + direction * d
            prev = j - direction
            if j < 0 or j >= len(win):
                break
            if win[j][0] <= 0 or win[prev][0] <= 0:
                break
            dx = win[j][1] - win[prev][1]
            dy = win[j][2] - win[prev][2]
            disp += math.sqrt(dx * dx + dy * dy)
            tri = (radius + 1 - d) / (radius + 1)
            gate = min(1.0, max(0.0, 1.0 - disp / tol))
            wj = tri * gate
            if wj <= 0:
                break
            w[j] = wj
    return w


# ------------------------------------------------------------ spatial bits
def _extreme(src, r, take_max):
    """Separable min/max over a (2r+1) square, edges clamped."""
    h, wd = src.shape
    xi = np.clip(np.arange(wd)[None, :] + np.arange(-r, r + 1)[:, None], 0, wd - 1)
    cols = src[:, xi]                                    # (h, 2r+1, w)
    tmp = (cols.max(1) if take_max else cols.min(1)).astype(np.float32)
    yi = np.clip(np.arange(h)[None, :] + np.arange(-r, r + 1)[:, None], 0, h - 1)
    rows = tmp[yi, :]                                    # (2r+1, h, w)
    return (rows.max(0) if take_max else rows.min(0)).astype(np.float32)


def dilate(a, r):
    return _extreme(a, r, True) if r > 0 else a


def erode(a, r):
    return _extreme(a, r, False) if r > 0 else a


def close(a, r):
    return erode(dilate(a, r), r)


def open_(a, r):
    return dilate(erode(a, r), r)


def blur121(src):
    """One separable [1 2 1]/4 pass, edges clamped, terms in the JS order."""
    h, wd = src.shape
    s = src.astype(np.float64)
    left = s[:, np.clip(np.arange(wd) - 1, 0, wd - 1)]
    right = s[:, np.clip(np.arange(wd) + 1, 0, wd - 1)]
    tmp = ((left + 2.0 * s + right) / 4.0).astype(np.float32).astype(np.float64)
    up = tmp[np.clip(np.arange(h) - 1, 0, h - 1), :]
    dn = tmp[np.clip(np.arange(h) + 1, 0, h - 1), :]
    return ((up + 2.0 * tmp + dn) / 4.0).astype(np.float32)


# ----------------------------------------------------------------- polish
def polish_frame(win, c, strength, st=None):
    """One frame. `win` is a contiguous list of float32 masks, `c` the centre."""
    p = params(strength)
    mid = win[c]
    if p['u'] <= 0:
        return mid.astype(np.float32, copy=True)
    cur = None
    if p['radius'] > 0 and len(win) > 1:
        S = st if st is not None else [stats(m) for m in win]
        wt = weights(S, c, p['radius'])
        total = float(wt.sum())
        if total > 0 and any(wt[j] > 0 for j in range(len(wt)) if j != c):
            acc = np.zeros(mid.shape, np.float32)
            for j, wj in enumerate(wt):
                if wj <= 0:
                    continue
                acc = (acc.astype(np.float64) + wj * win[j].astype(np.float64)) \
                    .astype(np.float32)
            cur = (acc.astype(np.float64) / total).astype(np.float32)
    if cur is None:
        cur = mid.astype(np.float32, copy=True)
    if p['morph'] > 0:
        cur = open_(close(cur, p['morph']), p['morph'])
    for _ in range(p['blur']):
        cur = blur121(cur)
    return cur


def polish_sequence(masks, strength):
    p = params(strength)
    if p['u'] <= 0:
        return [m.astype(np.float32, copy=True) for m in masks]
    st = [stats(m) for m in masks]
    out = []
    r = p['radius']
    for i in range(len(masks)):
        lo, hi = max(0, i - r), min(len(masks) - 1, i + r)
        out.append(polish_frame(masks[lo:hi + 1], i - lo, strength, st[lo:hi + 1]))
    return out


# ------------------------------------------------------------- directories
def _frames(d):
    return sorted(f for f in os.listdir(d) if f.lower().endswith('.png'))


def polish_dir(src, dst, strength, progress=None):
    """A directory of soft mask PNGs -> a directory of polished ones.

    Two passes and a ring buffer rather than the whole sequence in memory: a
    189-frame 720p subject is 700 MB as float32 and 4 MB as a window of seven.
    """
    files = _frames(src)
    if not files:
        raise RuntimeError('no mask PNGs in ' + src)
    p = params(strength)
    r = p['radius']
    os.makedirs(dst, exist_ok=True)

    st = []
    for f in files:
        st.append(stats(load_mask(os.path.join(src, f))))

    cache = {}

    def get(i):
        m = cache.get(i)
        if m is None:
            m = load_mask(os.path.join(src, files[i]))
            cache[i] = m
        return m

    for i, f in enumerate(files):
        lo, hi = max(0, i - r), min(len(files) - 1, i + r)
        win = [get(k) for k in range(lo, hi + 1)]
        save_mask(polish_frame(win, i - lo, strength, st[lo:hi + 1]),
                  os.path.join(dst, f))
        for k in list(cache):
            if k < lo:
                del cache[k]
        if progress:
            progress(i + 1, len(files))
    return {'frames': len(files), 'strength': int(strength), 'radius': r,
            'morph': p['morph'], 'blur': p['blur'], 'dir': dst}


def polished_dir(job_dir, oid, strength, rebuild=False):
    """The cached polished masks for one subject at one strength.

    jobs/<jid>/masks/<oid>       the tracker's own output, never touched
    jobs/<jid>/polish/<oid>/<s>  this, built once and reused by every render
    """
    src = os.path.join(job_dir, 'masks', str(oid))
    s = int(round(float(strength)))
    if s <= 0:
        return src, None
    dst = os.path.join(job_dir, 'polish', str(oid), str(s))
    done = os.path.join(dst, '.done')
    if os.path.exists(done) and not rebuild:
        return dst, None
    shutil.rmtree(dst, ignore_errors=True)
    info = polish_dir(src, dst, s)
    with open(done, 'w') as f:
        f.write(str(info['frames']))
    return dst, info


if __name__ == '__main__':
    import argparse, json, time
    ap = argparse.ArgumentParser(description='polish a directory of mask PNGs')
    ap.add_argument('--masks', required=True)
    ap.add_argument('--out', required=True)
    ap.add_argument('--strength', type=float, default=70)
    a = ap.parse_args()
    t0 = time.perf_counter()
    info = polish_dir(a.masks, a.out, a.strength)
    info['elapsed_s'] = round(time.perf_counter() - t0, 2)
    print(json.dumps(info, indent=2))
